format-common/src/main/java/io/cdap/plugin/format/plugin/AbstractFileSinkConfig.java

```diff
@@ -84,13 +84,32 @@ public void validate(FailureCollector collector) {
           .withConfigProperty(NAME_SUFFIX).withStacktrace(e.getStackTrace());
       }
     }
+
+    FileFormat format = FileFormat.CSV;
     try {
-      getFormat();
+      format = getFormat();
     } catch (IllegalArgumentException e) {
       collector.addFailure(e.getMessage(), null).withConfigProperty(NAME_FORMAT).withStacktrace(e.getStackTrace());
     }
     try {
-      getSchema();
+      Schema schema = getSchema();
+      // Checks if the output schema has fields that are not simple type.
+      // If there are non-simple field types, then serializing as CSV might not be appropriate.
+      // This restriction is on Csv, Delimited and Tab separated files.
+      if (format == FileFormat.CSV || format == FileFormat.DELIMITED || format == FileFormat.TSV) {
+        boolean simpleFields = true;
+        for (Schema.Field field : schema.getFields()) {
+          if (!field.getSchema().isSimpleOrNullableSimple()) {
+            simpleFields = false;
+            break;
+          }
+        }
+        collector.addFailure(
+          String.format("Input has multi-level structure that cannot be represented appropriately in %s.",
```

Comment:
should we use the word "nested" in the message? Also, it would help to add the name of the field that has the problem?

Reply:
I thought about that. But if there are multiple fields then we need to list
them all. But ultimately, if you think from user perspective we are not
suggesting them to remove the fields, the intent here is to suggest user to
use the right format for the data that they have. Suggesting to remove the
field can be conceived as limitation. So had to carefully word it.